### ISSUE TITLE
Managed Kubernetes in CAPI

### ISSUE THREAD

**Issue by richardcase** [kubernetes-sigs/cluster-api#7494]
/kind design
/kind api-change
/area api

**User Story**

As a cluster service consumer, I want to use Cluster API to provision and manage Kubernetes Clusters that utilize my service provider's Managed Kubernetes Service (i.e. EKS, AKS, GKE), So that I don’t have to worry about the management/provisioning of control plane nodes, and so I can take advantage of any value add services offered by the service provider.

**Detailed Description**

Let's take another look at how we managed Kubernetes services can be presented in CAPI and its providers.  

The original [Managed Kubernetes in CAPI proposal](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220725-managed-kubernetes.md), based on guidance from the community at the time, explicitly did not consider (i.e. out of scope) changes to CAPI itself and focused purely on fitting managed Kubernetes into the existing API types. The existing proposal is useful in the short term to provider guidance to provider implementers.

However, it has become apparent that the recommendations made in the proposal, whilst helpful, are less than ideal when it comes to managed Kubernetes.  Some areas where it's less than ideal (not an exhaustive list):

- It's assumed that the infra cluster provisions & reports the control plane endpoint. However, for managed k8s, an api server endpoint is usually created as part of creating the service in the cloud provider.
- Is there a difference between infra cluster and control plane when it comes to managed k8s? Arguably not, and so do we need both constructs?
- Some services have the need for multiple kubeconfigs

So, we should re-look at how we represent managed Kubernetes in CAPI and include the option of changes to CAPI itself to represent managed k8s as a first class citizen.

It's expected that this will result in a "Managed Kubernetes in CAPI v2" proposal.

**Anything else you would like to add:**

We should still consider the recommendations made in the original proposal (i.e. option 3), along with any new options that are a result of looking at changes to capi itself. We can then consider the trade-offs and decide a longer-term strategy for managed Kubernetes in capi.

Short-term, the 3 main cloud providers could achieve consistency quickly by going with option 2 from the proposal. Although not the recommended approach, it was still called out as an option. However, the Oracle provider implemented their managed service according to the document and went with option 3, so we need to consider this as well.

Some related issues that we can also consider as part of this work:

- #6520 
- #3661 
- #7032
- https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2739

/assign richardcase
/assign jackfrancis
/assign pydctw

**Comment by jackfrancis** [comment]
@richardcase thank you for this, very succinctly stated and with all the important items. 🙏 

I would like to call out how this proposal effects our "consistency across cloud providers" objective. You mention in the description several real-world issues that make consistency a challenge in the current landscape (in practice it would mean all cloud providers committing to either option 2 or 3 as defined in [current proposed set of recommendations](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220725-managed-kubernetes.md)).

I'd like to get some consensus from folks around the viability of pinning the "consistency across cloud providers" objective to the landing and eventual implementing of *this* new proposal instead of the [already landed proposal](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220725-managed-kubernetes.md). This would suggest a few obvious outcomes:

- in the near term, cloud providers implementing (or re-implementing) Managed Kubernetes would make a best effort to adhere to either option 2 or 3 [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220725-managed-kubernetes.md) according to provider-specific (and that provider's customers) criteria.
- consistency-as-a-goal means that this new proposal would accept a higher standard of cross-provider affimation as part of the specification gathering and finalization process: this new spec will in practice be an "API requirement" for implementing provider Managed Kubernetes in Cluster API, so we want to make an extra effort to involve the input of all providers in the ecosystem
- consistency-as-a-goal would be explicitly delayed for the Cluster API Managed Kubernetes community until this new proposal is accepted, implemented, and released (perhaps, e.g., as part of a `v1` API of Cluster API)
- we probably will also want to include in the scope of this effort some provider recommendations for migrating the current, existing Managed Kubernetes solutions onto the eventual Cluster API-blessed solution; by the time this consistent API is available to the community, we should anticipate that many customers will have already built platforms on top of the previous provider-specific Managed Kubernetes APIs

From the capz community: @mtougeron @zmalik @luthermonson @LochanRn @NovemberZulu (there are others as well), do you have any opinions on a willingness to adopt an eventual, Cluster API-specific Managed Kubernetes standard that will be consistent across cloud providers at a future date according to a new API? Does such a strategy have downsides to current platform adoption?

cc @CecileRobertMichon @nojnhuh

**Comment by pydctw** [comment]
Great write up, @richardcase. 

I also want to add what I heard from managed Kubernetes customers in the community.  As pointed out in @jackfrancis's [CAPZ Managed Kubernetes evolution proposal](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2739), the typical customer persona is a Cluster API, multi-cloud, Managed Kubernetes customer first of all and what they care about is consistency across Managed Kubernetes. They said they don't really expect consistency between managed and unmanaged Kubernetes in CAPI.

Considering managed Kubernetes can utilize many value added services offered by cloud providers, i.e. managed addons and easy installation mechanism, built-in health checks and scaling capabilities, I think managed Kubernetes proposal should focus on making it easy to bring in managed services' ever-increasing capabilities instead of trying to fit into current CAPI contracts that were designed for unmanaged Kubernetes.

**Comment by fabriziopandini** [comment]
> They said they don't really expect consistency between managed and unmanaged Kubernetes in CAPI.
> think managed Kubernetes proposal should focus on making it easy to bring in managed services' ever-increasing capabilities instead of trying to fit into current CAPI contracts that were designed for unmanaged Kubernetes.

Frankly speaking, I think we should be really careful in taking this path
Most of the value of Cluster API is having a common abstraction over different types of infrastructure, enabling higher-level tools to build truly multi-cluster/multi-infrastructure in a very effective way. 

So, in my humble opinion, this task should be about how to embrace managed Kubernetes while preserving a common abstraction, not about inventing a new abstraction.

**Comment by richardcase** [comment]
I agree @fabriziopandini 👍  The desire to keep managed / unmanaged as close as possible in the long term will be carefully considered. I think there are ways to do that whilst still being more implementation friendly to the managed provider implementors.

**Comment by jackfrancis** [comment]
The challenge here is that we have a trade-off in terms of *what* to prioritize for "close"-ness.

1. self-managed clusters and managed clusters on capi share a common API
  - in this scenario we will probably have to accept that that common API will _behave_ slightly differently between managed and unmanaged
2. preserve existing capi behavior for self-managed clusters
  - in this scenario we are probably nudged towards a dedicated API for managed (probably as simple as a single CRD, but still it would be edicated to managed k8s)

I'm not sure what the right answer is, looking forward to input from the community!

**Comment by shyamradhakrishnan** [comment]
@jackfrancis we need to understand the flaws of the current managed cluster CAPI proposal. It did tie nicely to CAPI API such as control plane, machine pool etc and also to Kubernetes constructs. We in OCI did not find any huge problems implementing using the proposal. Ofcourse there is a minor naming problem in that the control plane translates to the Cluster object in cloud provider, but technically the cluster(for example OKE/AKS) etc are really running managed control plane. 
If you did not have to migrate and were starting afresh, would you have been OK with the proposal? I would suggest we do think about that since we had spent quote a bit of time on the proposal, so anything new should have proper justification. Also another point to consider is can the current proposal be modified to answer all the concers.

Answering some questions above, do we need infra cluster and control plane, thats how current CAPI works and we can question that as well right? For example in current implementation, the infra cluster creates all the network infrastructure, the kubeadmcontrol plane creates the control plane nodes. We are doing the same thing with our impl, the infra cluster creates the network and any other infra, the control plane creates the managed control plane(which is called cluster by most cloud providers)

We can go through our impl and API in a meeting if required. I apologise if I did not understand the problem statement here or did nit explain properly.

**Comment by shyamradhakrishnan** [comment]
Sounds good @jackfrancis , a small change is that capa needs to be modified to support cluster class, as current implementation does not support it.

**Comment by puja108** [comment]
> There could be cases where an infra cluster could be used to provide something on top of "vanilla managed Kubernetes", people were quoting some examples about private clusters or additional security groups if I remember well

I mentioned something along those lines at KubeCon. My main point being that configurability options present with the managed Kubernetes solutions out there should also be possible with our CAPI solution for them for two reasons:
1. If our configurability is too low and limited to some generic cases, people will have good reasons not to use CAPI but go with solutions like the KubeCon NA Keynote from cruise, who built their own abstraction with Crossplane.
2. If coverage of options is too low, we will run into many blockers for future roadmap stories like adoption of existing managed Kubernetes clusters with CAPI.

**Comment by puja108** [comment]
Sorry Jack, this issue somehow got lost in my backlog.

I did not mean the comment with regard to any current or planned spec. My worries that I mentioned at KCSNA came more from talking to end users at KubeCon and in our virtual end user meetups, where I was seeing a curious rise in end users looking at Cluster API, maybe even trying it out, but then going for a self-built Crossplane-based solution in combination with Managed Control Planes like EKS and AKS. And the three end user companies I asked for the main reason behind this choice mentioned that they had special requirements for configurability or features that they would have had to implement around CAPI, e.g. private network support.

If this is irrelevant wrt the current proposal, then I might have gone off-topic for this thread.

**Comment by puja108** [comment]
Sadly I did not do any structure requirement gathering. I was mainly poking people at events and getting very short feedback. I think it would be worth it to gather them as a group in a more structured way.
As for the private network support, I think some this has been brought up at the CAPA (and soon CAPZ) level by some of my colleagues, and we internally have very specific needs from certain customers, which I hope we can slowly bring upstream, once the stressful PoC phases are over. I'll talk to the teams to see when and how we can bring things back upstream again.

**Comment by jackfrancis** [comment]
Thanks for weighing in @puja108.

At first glance it seems like these configuration requirements are provider-specific, but we'd love for you to join our regular feature group discussions around managed k8s in cluster api, your time permitting. Meeting details are in the markdown file in this PR that is en route to capi main branch:

https://github.com/kubernetes-sigs/cluster-api/pull/7546

**Comment by jackfrancis** [comment]
Just a quick status update that this CAEP is near to reaching definitional consensus:

- https://github.com/kubernetes-sigs/cluster-api/pull/8500

Following that an implementation effort will commence to carry out the work proposed above.

**Comment by jackfrancis** [comment]
Follow-up status update: last minute scope enhancement under consideration!

Specifically, @vincepri would like us to consider a new CRD definition for the control plane endpoint to aid any new behavioral flexibility we add to the existing assumptions of an infra provider's cluster and control plane resources. This may also help unblock forward progress on a long-standing request to support multiple control plane endpoints: https://github.com/kubernetes-sigs/cluster-api/issues/5295

See: https://kubernetes.slack.com/archives/C8TSNPY4T/p1686781200032249

**Comment by shyamradhakrishnan** [comment]
@jackfrancis this will benefit OKE and other managed provider as well I think. For example OKE support private and public endpoints https://docs.oracle.com/en-us/iaas/api/#/en/containerengine/20180222/datatypes/ClusterEndpoints , I am pretty sure other providers will also have this. Even in non managed cases, a Load balancer can have public and private endpoints. So theoretically, this will have a lot of benefits.

**Comment by sbueringer** [comment]
@jackfrancis Q: Is this issue done now that 8500 is merged? I assume we would want to track the implementation, maybe let's create a new umbrella issue for the implementation? (which would include the tasks from the PR description: https://github.com/kubernetes-sigs/cluster-api/pull/8500#issue-1661235911)
